postgres/datadog_checks/postgres/statement_samples.py

```diff
@@ -343,12 +344,14 @@ def _collect_plan_for_statement(self, row):
                 },
                 'postgres': {k: v for k, v in row.items() if k not in pg_stat_activity_sample_exclude_keys},
             }
+            event['timestamp'] = time.time() * 1000
             if row['state'] in {'idle', 'idle in transaction'}:
                 if row['state_change'] and row['query_start']:
                     event['duration'] = (row['state_change'] - row['query_start']).total_seconds() * 1e9
-                    event['timestamp'] = time.mktime(row['state_change'].timetuple()) * 1000
-            else:
-                event['timestamp'] = time.time() * 1000
+                    # If the transaction is idle then we have a more specific "end time" than the current time at
+                    # which we're collecting this event. Can only be done if the datetime has a timezone.
+                    if row['state_change'].tzinfo:
+                        event['timestamp'] = get_timestamp(row['state_change']) * 1000
```

Comment:
Just want to make sure I have all of the cases down:

1. `row['state_change']` is timezone-aware. Everything will work as expected.
2. `row['state_change']` is not timezone-aware, the agent is running on the database host.
3. `row['state_change']` is not timezone-aware, the agent is running off-host. 

As long as case #1 is guaranteed, we won't have any problems. #2 should also be fine (unless the database is using a different timezone than the host it's running on).

#3 would mean we default to using the agent timestamp. But because all events are aligned on UTC, we will always be close to the actual DB timestamp? Is this correct?

Reply:
According to [the docs](https://www.postgresql.org/docs/9.6/monitoring-stats.html) this is a `timestamp with timezone` so the timezone info should always be there. It would only be missing if there's a bug in `psycopg` or if the database host is misconfigured somehow. 

> #2 should also be fine (unless the database is using a different timezone than the host it's running on).
Yes in that case we use `now()` as the timestamp, which is still fine because if we're doing one collection per second we should be off by at most one second. 

> #3 would mean we default to using the agent timestamp. But because all events are aligned on UTC, we will always be close to the actual DB timestamp? Is this correct?
Yeah, same reasoning as `#2`, falling back to `now()` should still be within one second of the actual event time, unless the agent was down for a while then woke up and saw some stale connection that had finished its last query a long time ago.